src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
+
+    return canSeeHelper(X,Y, occludingObjects)
 ### Specifiers
 
 def With(prop, val):
-	"""The :grammar:`with <property> <value>` specifier.
+    """The :grammar:`with <property> <value>` specifier.
 
-	Specifies the given property, with no dependencies.
-	"""
-	return Specifier(prop, val)
+    Specifies the given property, with no dependencies.
+    """
+    return Specifier("With", {prop: 1}, {prop: val})
 
 def At(pos):
-	"""The :grammar:`at <vector>` specifier.
+    """The :grammar:`at <vector>` specifier.
 
-	Specifies :prop:`position`, with no dependencies."""
-	pos = toVector(pos, 'specifier "at X" with X not a vector')
-	return Specifier('position', pos)
+    Specifies :prop:`position`, with no dependencies.
+    """
+    pos = toVector(pos, 'specifier "at X" with X not a vector')
+    return Specifier("At", {'position': 1}, {'position': pos})
 
 def In(region):
-	"""The :grammar:`in/on <region>` specifier.
+    """The :grammar:`in <region>` specifier.
+
+    Specifies :prop:`position` and :prop:`parentOrientation`, with 
+    no dependencies.
+    """
+    region = toType(region, Region, 'specifier "in R" with R not a Region')
+    pos = Region.uniformPointIn(region)
+    props = {'position': 1}
+    values = {'position': pos}
+    if alwaysProvidesOrientation(region):
+        props['parentOrientation'] = 3
+        values['parentOrientation'] = region.orientation[pos]
+    return Specifier("In", props, values)
+
+def ContainedIn(region):
+    """The :grammar:`contained in <region>` specifier.
+
+    Specifies :prop:'position', :prop:`parentOrientation`, and :prop:'regionContainedIn',
+    with no dependencies.
+    """
+    region = toType(region, Region, 'specifier "contained in R" with R not a Region')
+    pos = Region.uniformPointIn(region)
+    props = {'position': 1, 'regionContainedIn': 1}
+    values = {'position': pos, 'regionContainedIn': region}
+    if alwaysProvidesOrientation(region):
+        props['parentOrientation'] = 3
+        values['parentOrientation'] = region.orientation[pos]
+    return Specifier("ContainedIn", props, values)
+
+
+def On(thing):
+    """The :specifier:`on {X}` specifier.
+
+    Specifies :prop:'position' and :prop:'parentOrientation', depending on
+    :prop:'onDirection', :prop:'baseOffset', :prop:'contactTolerance'.
+
+    Note that while :specifier:`on` can be used with `Region`, `Object` and `Vector`,
+    it cannot be used with a distribution containing anything other than `Region`.
+
+    May be used to modify an already specified :prop:'position' property if a compatible
+    specifier has already done so.
+
+    Allowed forms:
+        on <region>
+        on <object>
+        on <vector>
+    """
+    if isA(thing, Object):
+        target = toType(thing.onSurface, Region, 'Cannot coax occupiedSpace of Object to Region')
+    elif canCoerce(thing, Region):
+        target = toType(thing, Region, 'specifier "on R" with R not a Region')
```

Comment:
Need to clarify the error message -- `on` accepts several types other than Regions now, but when will this specific message will be seen?